Dockerfile

```diff
@@ -1,78 +1,138 @@
-FROM python:3.7.1-stretch
+# Dockerfile contains multiple levels of images--
+# First, a base image containing dependencies shared by all other images:
+#   - apt dependencies
+#   - rust
+#   - ntl
+#   - pbc
+#   - charm
+#   - base pip dependencies (cython & setup.py)
+# 
+# Thereafter, it add ever increasing levels of dependencies-- 
+#   - Test requirements (including doc requirements)
+#   - Dev requirements (including aws)
+#
+# In order to build and push this to dockerhub, run:
+# docker build . --target test-image --tag dsluiuc/honeybadgermpc-docker-base
+# docker push dsluiuc/honeybadgermpc-docker-base:latest
+
+
+FROM python:3.7.3-slim AS base-image
 
 # Allows for log messages to be immediately dumped to the 
 # stream instead of being buffered.
 ENV PYTHONUNBUFFERED    1 
 
-# Adds cargo to path for rust dependencies
-ENV PATH                /root/.cargo/bin:$PATH
-
 # Path variables needed for Charm
 ENV LIBRARY_PATH        /usr/local/lib
 ENV LD_LIBRARY_PATH     /usr/local/lib
 
-# Downloads rust and sets it up
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain nightly-2018-10-24
+# Make sh point to bash
+# This is being changed since it will avoid any errors in the `launch_mpc.sh` script
+# which relies on certain code that doesn't work in container's default shell.
+RUN ln -sf bash /bin/sh
 
 # Install apt dependencies
-RUN apt-get update && apt-get install -y \
-    vim \
-    tmux \
+RUN apt-get update && apt-get install -y --no-install-recommends \
     bison \
+    curl \
     flex \
+    g++ \
+    git \
+    libflint-dev \
     libgmp-dev \
     libmpc-dev \
     libmpfr-dev \
-    libflint-dev
+    libssl-dev \
+    make \
+    openssl \
+    tmux \
+    wget \
+    vim 
+
+# This is needed otherwise the build for the power sum solver will fail.
+# This is a known issue in the version of libflint-dev in apt.
+# https://github.com/wbhart/flint2/issues/217
+# This has been fixed if we pull the latest code from the repo. However, we want
+# to avoid compiling the lib from the source since it adds 20 minutes to the build.
+RUN sed -i '30c #include "flint/flint.h"' /usr/include/flint/flintxx/flint_classes.h
+RUN echo "alias cls=\"clear && printf '\e[3J'\"" >> ~/.bashrc
+
+# RUN python -m venv /opt/venv
+# ENV PATH "/opt/venv/bin:${PATH}"
+
+# Downloads rust and sets it up
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain nightly-2018-10-24
+ENV PATH "/root/.cargo/bin:${PATH}"
+
 # Download and build NTL from source
 # Shoup recommends not using O3
-RUN wget -qO- https://www.shoup.net/ntl/ntl-11.3.2.tar.gz | tar xzvf - \
-         &&  cd ntl-11.3.2/src \
-         && ./configure CXXFLAGS="-g -O2 -fPIC -march=native -pthread -std=c++11" \
-         && make -j4 \
-         && make install
+RUN curl -so - https://www.shoup.net/ntl/ntl-11.3.2.tar.gz | tar xzvf -
+WORKDIR /ntl-11.3.2/src  
```

Comment:
Doesn't each line act as a new layer? Why are we going back to what we had earlier?

Reply:
So, mostly this is more development friendly. I've been going through the docs, and having many layers isn't really problematic--

1. It can cache up until the first thing that invalidates the cache, so builds are faster
2. Modern docker isn't impacted much from it performance / memory wise.
3. Easier to debug the dockerfile.